### ISSUE TITLE
GLTFLoader display model is darker

### ISSUE THREAD

**Issue by Maplesog** [mrdoob/three.js#12554]
##### Description of the problem 

My modal is .obj, and it displays well when I use objLoader and mtlLoader . I use obj2gltf to convert my obj modal to gltf. And then it displays more darker with GLTFLoader. what's the problem?

##### Three.js version

- [ ] Dev
- [x] r87
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

**Comment by takahirox** [comment]
Can other viewers/libs render as you expect? If not, it'd be an obj2gltf issue.

**Comment by Mugen87** [comment]
You can also upload your model to [clara.io](https://clara.io/) or [sketchfab](https://sketchfab.com) and compare the results. In this way, you don't have to write any code.

**Comment by donmccurdy** [comment]
Assuming you're converting to glTF 2.0, obj2gltf is converting to a metal-rough PBR workflow, i.e. THREE.MeshStandardMaterial. This may give different results, and/or look better with an environment map.

As others mention, comparing against various viewers (https://gltf-viewer.donmccurdy.com/, http://sandbox.babylonjs.com/) or including your source and output models would make it easier to help you. It may be an obj2gltf issue.

**Comment by rck1** [comment]
I am experiencing the same issue.  I am not using  obj2gltf.  If you look at the gltf loader example (https://threejs.org/examples/#webgl_loader_gltf) with GLTFLoader.js r87 vs r88, you will see that the r88 is significantly darker.

**Comment by donmccurdy** [comment]
@rck1 that is another issue, see https://github.com/donmccurdy/three-gltf-viewer/issues/51. r88 is correct, THREE.GLTFLoader <=r87 had used incorrect texture encoding.

**Comment by rck1** [comment]
@donmccurdy, thanks for the link and explanation.  That makes sense but I was hoping there would be  consistency across the image applications that I use and the results that I'm seeing with the GLTFLoader.  For instance, looking at the texture in a program like Photoshop and rendered using three.js/GLTFLoader now looks very different.  With r87 it looks the same.  But the results using JSONLoader (r87 or r88) is still consistent.

**Comment by donmccurdy** [comment]
@rck1 Base color and emissive textures in glTF are expected to be in [sRGB colorspace](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#pbrmetallicroughnessbasecolortexture), not linear. If you're creating textures in Photoshop, you will want to set export options accordingly. glTF-specific tools, like the Blender exporter, should automatically do this conversion (and it would be a bug if they are not).

**Comment by rck1** [comment]
@donmccurdy I'll double check the color management workflow across all my applications to ensure sRGB colorspace.  Thanks for your help with this.

**Comment by Maplesog** [comment]
@donmccurdy Thanks for your suggest.The following modal is converted by obj2gltf and show with threejs GLTF viewer. Obviously it's more darker than normal and not expected. And I got some suggests yesterday  to return THREE.MeshLambertMaterial in  threejs GLTFLoader.js souce code line 299. The modal seemed more lighter than before but still dark.
<img width="1642" alt="qq20171103-164541 2x" src="https://user-images.githubusercontent.com/18155748/32366307-8ee2272c-c04b-11e7-94c7-ebc9cf91cdb2.png">

So I changed the viewer http://sandbox.babylonjs.com/ you told me.Finally I find what I expected.
<img width="1628" alt="qq20171103-164730 2x" src="https://user-images.githubusercontent.com/18155748/32366498-57725310-c04c-11e7-9e67-287dc737fa65.png">

What should I do with  GLTFLoader.js of  threejs or is there any way threejs  learning from babylonjs?

**Comment by rck1** [comment]
@donmccurdy  I did some digging and think I gained some surface-level knowledge on this topic of color management.  I think three.js utilizes linear color space throughout it's pipeline for lighting calculations, etc.  So when you flag the map/texture as sRGB, it removes the gamma correction so calculations can correctly be done in linear color space.  I think this is the right approach and is followed by most standard 3D applications.  However, by default, three.js doesn't gamma correct it's rendering output by default.  So removing the gamma correction in sRGB texture results in a darker than intended output. 

@Maplesog Try adding the following to your renderer. 

`renderer.gammaFactor = 2.2;`
`renderer.gammaOutput = true;`

This should perform the proper gamma correction.

**Comment by donmccurdy** [comment]
> What should I do with GLTFLoader.js of threejs or is there any way threejs learning from babylonjs?

@Maplesog it's probably not a threejs problem: the lighting in my viewer is not great, and generally looks too dark for many models. That's just a matter of adjusting the scene lighting, though... You can tweak knobs in the settings, or set up your own scene to make it as bright as you want. I've been meaning to improve the defaults in my viewer, but haven't found something I'm happy with: https://github.com/donmccurdy/three-gltf-viewer/issues/44

Swapping in MeshLambertMaterial will certainly change what you see, but is technically not a shading model supported by the glTF format (currently) so you'd need to do those modifications yourself.

If you're testing in your own scene, make sure to include an `envMap`. MeshStandardMaterial tends to look better with that.

Finally, if you can share the model this type conversation is often more productive. If you've only tested the model in my viewer, it might be better to [file an issue there](https://github.com/donmccurdy/three-gltf-viewer/issues)... lighting is not dictated by GLTFLoader or three.js, it's up to the viewer application to choose that.

@rck1 thanks, this sounds like it's worth looking into. If you have an example you can share, that would be helpful.

**Comment by Maplesog** [comment]
@rck1 It works! @donmccurdy Your suggestion let me have a deeper understanding of the GLTFloader, Thank you very much!

**Comment by donmccurdy** [comment]
Reopening this, as I do think there may be an issue. I've put more lighting knobs in my viewer, https://gltf-viewer.donmccurdy.com/, with what I hope are reasonable defaults, to help sort this out.

With all default settings (`directIntensity=0.8`, `ambientIntensity=0.3`), here is a comparison of texture encodings.

| sRGB | Linear |
|---|---|
| ![srgb](https://user-images.githubusercontent.com/1848368/33236417-a31b8d32-d207-11e7-9dec-3f3395c9a546.png) | ![linear](https://user-images.githubusercontent.com/1848368/33236418-a4b18e12-d207-11e7-935f-6e50fe48de2d.png) |

This example is pretty representative of the [glTF-Sample-Models catalog](https://github.com/KhronosGroup/glTF-Sample-Models/) — they all look better lit, and maybe a little washed out, with THREE.LinearEncoding left as the default. Per [glTF 2.0 spec](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#pbrmetallicroughnessbasecolortexture), these textures are supposed to be sRGB. And with default lighting in several other viewers, the same model looks reasonable:

| viewer | preview |
|---|---|
| three.js | ![srgb](https://user-images.githubusercontent.com/1848368/33236522-534ec7fc-d20b-11e7-82ad-a9e2c986dda4.png) |
| babylon.js | *broken?* |
| qtek | ![screen shot 2017-11-25 at 6 05 36 pm](https://user-images.githubusercontent.com/1848368/33236521-4c136970-d20b-11e7-9680-d075b76b72ee.png) |
| sketchfab | ![screen shot 2017-11-25 at 6 07 12 pm](https://user-images.githubusercontent.com/1848368/33236536-85ff9fa0-d20b-11e7-9db6-bd1df102cdae.png) |

Original: https://sketchfab.com/models/f2f13a8630004b1c82730d8b9ffa0e1f

If I leave the default THREE.LinearEncoding on, our result looks quite close to Sketchfab. So... what's going on here? 😕

**Comment by mrdoob** [comment]
https://twitter.com/mrdoob/status/934570830016626688

**Comment by spiraloid** [comment]
no more gamma?

**Comment by Mugen87** [comment]
`WebGLRenderer.gammaInput` and `WebGLRenderer.gammaOutput` have been removed with `R112`. This is mentioned in the migration guide:

https://github.com/mrdoob/three.js/wiki/Migration-Guide#r111--r112

Instead of doing this:
```js
renderer.gammaFactor = 2.2;
renderer.gammaOutput = true;
```
do this:
```js
renderer.outputEncoding = THREE.sRGBEncoding;
```

**Comment by acmoles** [comment]
I'm finding my textures display darker despite setting 
`renderer.outputEncoding = THREE.sRGBEncoding;`
They were correct in r111. 

I've also tried
`mesh.material.metalness = 0;`
`renderer.physicallyCorrectLights = true;`
Without success.

**Comment by Mugen87** [comment]
Please ask at the [forum](https://discourse.threejs.org/) for help. It's also recommended to share a live example that demonstrates your issue.

**Comment by spelman7** [comment]
I solved a related issue today, and thought I'd post the solution here in case anyone needed it.

In the process of updating our three.js version, our loaded GLTFs got much, much darker. I tracked this back to the [r136 update](https://github.com/mrdoob/three.js/wiki/Migration-Guide#135--136) where gamma-related controls were removed from `THREE.WebGLRenderer`. 

I use a custom `ShaderMaterial` to render the GLTFs, and so I manually gamma-corrected in the fragment shader itself, for which [there is a jsm example](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/shaders/GammaCorrectionShader.js).
